crates/wallet/src/wallet/persisted.rs

```diff
@@ -1,130 +1,320 @@
+/// used by [`PersistedWallet`] (a light wrapper around [`Wallet`]) which enforces some level of
+/// safety. Refer to [`PersistedWallet`] for more about the safety checks.
+pub trait WalletPersister {
+    /// Error type of the persister.
+    type Error;
+
+    /// Initialize the `persister` and load all data.
+    ///
+    /// This is called by [`PersistedWallet::create`] and [`PersistedWallet::load`] to ensure
+    /// the [`WalletPersister`] is initialized and returns all data in the `persister`.
+    ///
+    /// # Implementation Details
+    ///
+    /// The database schema of the `persister` (if any), should be initialized and migrated here.
+    ///
+    /// The implementation must return all data currently stored in the `persister`. If there is no
+    /// data, return an empty changeset (using [`ChangeSet::default()`]).
+    ///
+    /// Error should only occur on database failure. Multiple calls to `initialize` should not
+    /// error. Calling [`persist`] before calling `initialize` should not error either.
+    ///
+    /// [`persist`]: WalletPersister::persist
+    fn initialize(persister: &mut Self) -> Result<ChangeSet, Self::Error>;
+
+    /// Persist the given `changeset` to the `persister`.
+    ///
+    /// This method can fail if the `persister` is not [`initialize`]d.
+    ///
+    /// [`initialize`]: WalletPersister::initialize
+    fn persist(persister: &mut Self, changeset: &ChangeSet) -> Result<(), Self::Error>;
+}
+
+type FutureResult<'a, T, E> = Pin<Box<dyn Future<Output = Result<T, E>> + Send + 'a>>;
+
+/// Async trait that persists [`PersistedWallet`].
+///
+/// For a blocking version, use [`WalletPersister`].
+///
+/// Associated functions of this trait should not be called directly, and the trait is designed so
+/// that associated functions are hard to find (since they are not methods!). [`WalletPersister`] is
+/// used by [`PersistedWallet`] (a light wrapper around [`Wallet`]) which enforces some level of
+/// safety. Refer to [`PersistedWallet`] for more about the safety checks.
+pub trait AsyncWalletPersister {
+    /// Error type of the persister.
+    type Error;
+
+    /// Initialize the `persister` and load all data.
+    ///
+    /// This is called by [`PersistedWallet::create_async`] and [`PersistedWallet::load_async`] to
+    /// ensure the [`WalletPersister`] is initialized and returns all data in the `persister`.
+    ///
+    /// # Implementation Details
+    ///
+    /// The database schema of the `persister` (if any), should be initialized and migrated here.
+    ///
+    /// The implementation must return all data currently stored in the `persister`. If there is no
+    /// data, return an empty changeset (using [`ChangeSet::default()`]).
+    ///
+    /// Error should only occur on database failure. Multiple calls to `initialize` should not
+    /// error. Calling [`persist`] before calling `initialize` should not error either.
```

Comment:
Same comment as above.
```suggestion
    /// error. Calling [`persist`] before calling `initialize` can error.
```